### PR TITLE
[FEATURE] Some terminal functionality that will be needed for the argument parser help page printing.

### PR DESCRIPTION
**Pull request by smehringer**
Requirement for #88 to work correctly. Must be merged before #88.

### CODE REVIEW

File: include/seqan3/core/detail/terminal.hpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+#pragma once
+
+#include <cstdio>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+#include <seqan3/core/platform.hpp>
+
+namespace seqan3::detail {
```
Comment (marehr) [inline include/seqan3/core/detail/terminal.hpp]: brace `{` on new line

File: include/seqan3/core/detail/terminal.hpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+/*!\file
+ * \author Svenja Mehringer <svenja.mehringer AT fu-berlin.de>
+ * \brief Transfered from seqan2. Used in the argument parser to check console properties.
+ */
+
+#pragma once
+
+#include <cstdio>
+#include <sys/ioctl.h>
+#include <unistd.h>
```
Comment (marehr) [inline include/seqan3/core/detail/terminal.hpp]: https://github.com/seqan/seqan3/wiki/File-structure-and-naming:
(an empty line between each block, sorted alphabetically within)

File: include/seqan3/core/detail/terminal.hpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+namespace seqan3::detail {
+
+// ----------------------------------------------------------------------------
+// Function is_terminal()
+// ----------------------------------------------------------------------------
+
+/*!\brief Check whether we are printing to a terminal.
+ *
+ * \returns bool True if code is run in a terminal, False otherwise.
```
Comment (marehr) [inline include/seqan3/core/detail/terminal.hpp]: we use mostly `\return`

File: include/seqan3/core/detail/terminal.hpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    return isatty(fileno(stdout));
+}
+
+// ----------------------------------------------------------------------------
+// Function get_terminal_size()
+// ----------------------------------------------------------------------------
+
+/*!\brief  Retrieve size of terminal.
+ *
+ * @signature bool get_terminal_size(cols, rows);
```
Comment (marehr) [inline include/seqan3/core/detail/terminal.hpp]: `\signature`? Hmm wrong documentation?

File: include/seqan3/core/detail/terminal.hpp [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+ *
+ * \returns A pair [rows, cols] for the number of row and column bytes of the
```
Comment (marehr) [inline include/seqan3/core/detail/terminal.hpp]: `\return`?

**Comment by marehr** [comment]
Just a question, is there any way to test this?

**Comment by smehringer** [comment]
I don't think so since this depends on the build environment.

File: include/seqan3/core/detail/terminal.hpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \author Svenja Mehringer <svenja.mehringer AT fu-berlin.de>
+ * \brief Transfered from seqan2. Used in the argument parser to check console properties.
```
Comment (h-2) [inline include/seqan3/core/detail/terminal.hpp]: Checks if program is run interactively and retrieves dimensions of terminal.

File: include/seqan3/core/detail/terminal.hpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+namespace seqan3::detail
+{
+
+// ----------------------------------------------------------------------------
+// Function is_terminal()
+// ----------------------------------------------------------------------------
+
+/*!\brief Check whether we are printing to a terminal.
+ *
+ * \return bool True if code is run in a terminal, False otherwise.
```
Comment (h-2) [inline include/seqan3/core/detail/terminal.hpp]: remove "bool" make true and false lower case.

File: include/seqan3/core/detail/terminal.hpp [hunk 8]
```diff
@@ -0,0 +1,86 @@
+// ============================================================================
+//                 SeqAn - The Library for Sequence Analysis
+// ============================================================================
+//
+// Copyright (c) 2006-2017, Knut Reinert & Freie Universitaet Berlin
+// Copyright (c) 2016-2017, Knut Reinert & MPI Molekulare Genetik
```
Comment (joergi-w) [inline include/seqan3/core/detail/terminal.hpp]: year 2018

File: include/seqan3/core/detail/terminal.hpp (hunk 6 above)
Comment (joergi-w) [reply include/seqan3/core/detail/terminal.hpp]: transfered => transferred

Comment (joergi-w) [reply include/seqan3/core/detail/terminal.hpp]: please be consistent: console or terminal

File: include/seqan3/core/detail/terminal.hpp [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+{
+
+// ----------------------------------------------------------------------------
+// Function is_terminal()
+// ----------------------------------------------------------------------------
+
+/*!\brief Check whether we are printing to a terminal.
+ *
+ * \return bool True if code is run in a terminal, False otherwise.
+ *               (Always false in Windows)
```
Comment (joergi-w) [inline include/seqan3/core/detail/terminal.hpp]: if you mean the Windows OS: isn't it *on* Windows? And maybe add the word *platform* or *OS* to make it clearer.

File: include/seqan3/core/detail/terminal.hpp [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+// ----------------------------------------------------------------------------
+// Function get_terminal_size()
+// ----------------------------------------------------------------------------
+
+/*!\brief  Retrieve size of terminal.
+ *
+ * \return A pair [rows, cols] for the number of row and column bytes of the
+ *          current terminal.
```
Comment (joergi-w) [inline include/seqan3/core/detail/terminal.hpp]: What means row and column bytes? Is it the number of...?

Comment (joergi-w) [reply include/seqan3/core/detail/terminal.hpp]: Your function only returns the number of columns (=width). So this is misleading documentation.

File: include/seqan3/core/detail/terminal.hpp [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    struct winsize w;
+    w.ws_row = 0;
+    w.ws_col = 0;
+
+    ioctl(STDOUT_FILENO, TIOCGWINSZ, &w);
+
+    return w.ws_col;
+}
+
+}  // namespace seqan::detail
```
Comment (joergi-w) [inline include/seqan3/core/detail/terminal.hpp]: newline at end of file

File: include/seqan3/core/detail/terminal.hpp [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+ * \return A pair [rows, cols] for the number of row and column bytes of the
+ *          current terminal.
+ */
+inline unsigned get_terminal_width()
+{
+    struct winsize w;
+    w.ws_row = 0;
+    w.ws_col = 0;
+
+    ioctl(STDOUT_FILENO, TIOCGWINSZ, &w);
```
Comment (joergi-w) [inline include/seqan3/core/detail/terminal.hpp]: I would like a comment what this does, as there are names that do not allow guessing (e.g. `TIOCGWINSZ`).

Comment (smehringer) [reply include/seqan3/core/detail/terminal.hpp]: See docu

File: include/seqan3/core/detail/terminal.hpp (hunk 11 above)
Comment (smehringer) [reply include/seqan3/core/detail/terminal.hpp]: done

**Pull request merged**